File: src/reader/utils/pc-utils.ts
import { DynamicBuffer } from './dynamic-buffer.js';
import { bytesToHex } from './hex-utils.js';

/**
 * Protocol Control (PC) word utility functions.
 *
 * Replaces Python `pcUtils.py`. Used for computing PC values and
 * assembling EPC / GB data payloads.
 */

/**
 * Compute the PC word-count length from a hex value string.
 * @param value Hex string of the data content.
 * @returns Number of 16-bit words required, or `undefined` if value is empty.
 */
export function getPcLen(value: string): number | undefined {
  if (!value) {
    return undefined;
  }
  const t1 = Math.floor(value.length / 4);
  const t2 = value.length % 4;
  if (t2 === 0) {
    return t1;
  }
  return t1 + 1;
}

/**
 * Compute the EPC PC value as a 4-character hex string.
 * @param value Hex string of the EPC content.
 * @returns 4-character hex string representing the PC word.
 */
export function getPcValue(value: string): string {
  const pcLen = getPcLen(value);
  if (pcLen === undefined) {
    return '';
  }
  const iPc = pcLen << 11;
  const buffer = new DynamicBuffer();
  buffer.putUint32BE(iPc);
  buffer.pos = 16;
  const bytes = buffer.readByteArray(16);
  return bytes ? bytesToHex(bytes) : '';
}

/**
 * Build the full EPC write payload: PC value + content padded to word boundary.
 * @param value Hex string of the EPC content.
 * @returns Hex string of PC + padded content.
 */
export function getEpcData(value: string): string {
  const pcLen = getPcLen(value);
  if (pcLen === undefined) {
    return '';
  }
  const valueLen = pcLen * 4;
  const pcValue = getPcValue(value);
  return pcValue + value.padEnd(valueLen, '0');
}

/**
 * Compute the GB (国标) PC value as a 4-character hex string.
 * @param value Hex string of the GB content.
 * @returns 4-character hex string representing the GB PC word.
 */
export function getGbPcValue(value: string): string {
  const pcLen = getPcLen(value);
  if (pcLen === undefined) {
    return '';
  }
  const iPc = pcLen << 8;
  const buffer = new DynamicBuffer();
  buffer.putUint32BE(iPc);
  buffer.pos = 16;
  const bytes = buffer.readByteArray(16);
  return bytes ? bytesToHex(bytes) : '';
}

/**
 * Build the full GB write payload: PC value + content padded to word boundary.
 * @param value Hex string of the GB content.
 * @returns Hex string of PC + padded content.
 */
export function getGbData(value: string): string {
  const pcLen = getPcLen(value);
  if (pcLen === undefined) {
    return '';
  }
  const valueLen = pcLen * 4;
  const pcValue = getGbPcValue(value);
  return pcValue + value.padEnd(valueLen, '0');
}

File: python/reader/utils/pcUtils.py
from uhf.reader.utils.HexUtils import bytesToHex
from uhf.reader.utils.byteBuffer import *


def getPcLen(value: str):
    """
    :rtype: int
    """
    if value:
        t1, t2 = divmod(len(value), 4)
        if t2 == 0:
            return t1
        elif t2 > 0:
            return t1 + 1


def getPcValue(value: str):
    """
    计算epc的pc值
    :param value: 16进制写入内容
    :return: 16进制 pc值
    """
    pcLen = getPcLen(value)
    iPc = pcLen << 11
    buffer = DynamicBuffer()
    buffer.putLong(iPc)
    buffer.pos = 16
    return bytesToHex(buffer.readBytes(16))


def getEpcData(value: str):
    """
    写6C epc 快捷返回pc+内容
    :param value: 16进制写入内容
    :return:pc+内容
    """
    value_len = getPcLen(value) * 4
    pc_value = getPcValue(value)
    return pc_value + value.ljust(value_len, "0")


def getGbPcValue(value: str):
    """
    计算GB 编码区的pc值
    :param value: 16进制写入内容
    :return: 16进制 pc值
    """
    pcLen = getPcLen(value)
    iPc = pcLen << 8
    buffer = DynamicBuffer()
    buffer.putLong(iPc)
    buffer.pos = 16
    return bytesToHex(buffer.readBytes(16))


def getGbData(value: str):
    """
    写GB 编码区 快捷返回pc+内容
    :param value: 16写入进制内容
    :return: pc+内容
    """
    value_len = getPcLen(value) * 4
    pc_value = getGbPcValue(value)
    return pc_value + value.ljust(value_len, "0")


if __name__ == '__main__':

    print(divmod(632,64))
    for i in range(1):
        print(i)
    l = [1, 2, 3, 4, 5]
    l.extend([4, 5, 6, 7])
    print(l)

